plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -419,37 +423,166 @@ public void downloadPlugins(List<Plugin> plugins) {
+        }
+
+        Map<String, List<Plugin>> allPlugins = new HashMap<>();
+        List<Plugin> pluginList = new ArrayList<>();
+        pluginList.add(plugin);
+
+        allPlugins.put(plugin.getName(), pluginList);
+        for (Plugin p : plugin.getDirectDependencies()) {
+            Map<String, Plugin> optimizedChildDependencies = findOptimalDependencies(p);
+            addToAllDependentPlugins(optimizedChildDependencies, allPlugins);
+        }
+
+        Map<String, Plugin> optimalDependencies = optimize(allPlugins);
+        plugin.setRecursiveDependencies(optimalDependencies);
+        return optimalDependencies;
+    }
+
+    /**
+     * Given all of the optimized dependencies of a particular child plugin, adds them to the map of all the optimized
+     * children dependencies. If a plugin with the same name occurs twice, it is added to a list of all dependent
+     * plugins with the same name
+     *
+     * @param optimizedChildDependencies map of entries corresponding to plugin names and plugin pairs
+     * @param allPlugins                 map of entries corresponding to plugin name and list of plugins that share that name pairs
+     */
+    public void addToAllDependentPlugins(Map<String, Plugin> optimizedChildDependencies,
+                                         Map<String, List<Plugin>> allPlugins) {
+        for (Map.Entry<String, Plugin> entry : optimizedChildDependencies.entrySet()) {
+            String dependencyName = entry.getKey();
+            if (!allPlugins.containsKey(dependencyName)) {
+                List<Plugin> samePlugins = new ArrayList<>();
+                samePlugins.add(entry.getValue());
+                allPlugins.put(dependencyName, samePlugins);
+            } else {
+                allPlugins.get(dependencyName).add(entry.getValue());
+            }
+        }
+    }
+
+    /**
+     * If there are any duplicated plugins in the list of all optimized plugins, removes them and their dependencies
+     *
+     * @param childDependencies map of all optimized dependencies, including duplicates
+     * @return optimized dependencies, which is a map of plugin name, plugin pairs
+     */
+    public Map<String, Plugin> optimize(Map<String, List<Plugin>> childDependencies) {
+        List<Plugin> duplicates = findDuplicates(childDependencies);
+
+        for (Plugin duplicate : duplicates) {
+            removeFromOptimalDependencies(duplicate, childDependencies);
+            List<Plugin> removedPluginsDependencies = new ArrayList<>(duplicate.getRecursiveDependencies().values());
+            for (Plugin dependency : removedPluginsDependencies) {
+                removeFromOptimalDependencies(dependency, childDependencies);
+            }
+        }
+
+        Map<String, Plugin> optimizedDependencies = new HashMap<>();
+
+        //after duplicates have been removed, should only be one plugin in the list
+        for (Map.Entry<String, List<Plugin>> entry : childDependencies.entrySet()) {
```

Comment:
I'd also use streams here:
```
Map<String, Plugin> optimizedDependencies = childDependencies.entrySet().stream()
    .filter(x -> x.getValue().isEmpty())
    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
If using some previous optimizations I suggested, you might not even need to run this (ie: removing based on values)

Reply:
instead `stream()` in this case always use `parallelStream()`